TRUNK-5400 - Add Orders.fulfillerStatus and Orders.fulfillerComment

fruether opened the pull request:
Resolving TRUNK-5400 - Add Orders.fulfillerStatus and Orders.fulfillerComment
https://issues.openmrs.org/browse/TRUNK-5400
## Description of what I changed

- Added the attributes to the Order object
- Made the attributes changeable 
- Added set() and getter() and made sure clone, and so on is still working
- Added a public method that allowed to change the status and comment of an order. This method was also added to the interface
- Added a test case to make sure all changes worked


## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->
see https://issues.openmrs.org/browse/TRUNK-5400
and for the discussion on https://talk.openmrs.org/t/adding-order-status-to-orders/17928/31

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

fruether commented:
@coveralls decreased but not because of the files that I changed. What should I do about this?

mogoodrich commented:
@fruether you will also need to create a liquibase changeset (in "liquibase-update-to-2.2.xml") to add these two columns to the Order table and update the Order.hbm.xml file.  Though I don't know how your test would pass without these... maybe I am missing something?

fruether commented:
@mogoodrich I need some advice.

When I add into Orders.hbm.xml the following lines of code:
```

<property name="fulfillStatus" column="fulfiller_status" length="50">
			<type name="org.hibernate.type.EnumType">
				<param name="enumClass">org.openmrs.Order$FulfillStatus</param>
				<param name="type">12</param>
			</type>
</property>
```

then the tests stop working and throw a bunch of exceptions. That was what I was having in liquidbase change:

	```
	<addColumn tableName="orders">
			<column name="fulfiller_status" type="varchar(50)" >
				<constraints nullable="true" />
			</column>
```

Do you have an idea why this is the case?

fruether commented:
@mogoodrich I updated the changes you mentioned. Do I need a databaseUpdate test?

mogoodrich commented:
@fruether  are you still having the issue with Orders.hbm.xml?  If so, let me know what the errors are.

To your other question, we don't have a good way to do database update tests, unfortunately.

fruether commented:
@mogoodrich I fixed the issue. The error was that the setter() was named differently. I called it setFulfill_er_Status instead of setFulfillStatus which is expected by some internal calls of Hibernate. As a result test cases failed. I fixed the issue and all test cases passed.

If there is no other was to test for the future if the elements are mapped to the databse and the database tables is changed, then I think it could be merged. Only open question would may be if I match the naming conventions for the DB? And if the variable names  (both in DB and Java) are good enough?

mogoodrich commented:
Thanks @fruether ... I made one comment on the ticket, but, also, I would name the getter/setter "fufillerStatus" instead of "fulfillStatus". This is the "more proper" name.  Thanks!

fruether commented:
@mogoodrich with what arguments to execute it? I was using 

`mvn` openmrs-sdk:setup -DserverId=webservices-dev -Dplatform=2.0.0
--
I could imagine Dplatform should be 2.2.0 but then I could not find the according files. I was further following https://wiki.openmrs.org/display/docs/OpenMRS+SDK+Step+By+Step+Tutorials#OpenMRSSDKStepByStepTutorials-Settingupdevenvironment

Is this sufficient or should I also execute **"Developing with the Reference Application"**
"" 
Sry for all the questions. Never used it before

mogoodrich commented:
No worries @fruether... is there another way you usually run OpenMRS? It doesn't necessarily need to be via the SDK but you should test starting up OpenMRS in some way.

fyi, I will not be working tomorrow and offline this weekend so likely will not be able to follow up again until next week.

fruether commented:
@mogoodrich I hope you had a nice weekend and day off!

I was running the setup of Open:MRS and afterwards executed a select statement on the orders database. It seems like it was executed and the database is changed accordingly:

<img width="563" alt="bildschirmfoto 2018-08-04 um 20 04 38" src="https://user-images.githubusercontent.com/1591040/43679212-e4c68b66-9821-11e8-9e47-f36955b2d9a6.png">
<img width="617" alt="bildschirmfoto 2018-08-04 um 20 07 16" src="https://user-images.githubusercontent.com/1591040/43679224-0480e1ae-9822-11e8-90c5-880572e99826.png">


Do you want me to perform more kind of proofs? Creating an order would be difficult without the REST service being implemented? Any kind of guidance would be appreciated since I am still new to open MRS? 

Thank you!

mogoodrich commented:
@fruether ... that should be enough, thanks @fruether ... I should be able to review and merge this week.

fruether commented:
@mogoodrich Thanks! You are welcome!

How about the REST service part of the ticket? Is is already assingned to a person or should I try to take it on as well?

wluyima commented on api/src/main/java/org/openmrs/Order.java:
```diff
@@ -115,6 +125,19 @@
+	 * @see org.openmrs.Order.FulfillStatus
+	 * @since 2.9.0
+	 */
+	private FulfillStatus fulfillerStatus = FulfillStatus.RECEIVED;
```
Do you need a default value? I'd imagine the status would be unknown upon placement of the order

mogoodrich replied:
+1.  Default should be null.

wluyima commented on api/src/main/java/org/openmrs/Order.java:
```diff
@@ -843,4 +871,21 @@ public Double getSortWeight() {
 	public void setSortWeight(Double sortWeight) {
 		this.sortWeight = sortWeight;
 	}
+	
```
Actually the @ since should be on these getters/setters since they are public and not the fields which are private

mogoodrich replied:
Looks like they are on these setters, am I missing something?  I'm fine if it's one the private fields as well for redundancy.

wluyima replied:
I was just saying they are redundant on a field because they are private, may be they can be useful for developers new to the project to quickly know when the properties were introduced by just glancing at the fields.

wluyima commented on api/src/main/java/org/openmrs/api/OrderService.java:
```diff
@@ -265,6 +265,20 @@
+	 * @should save the changed order
+	 */
+	@Authorized(PrivilegeConstants.EDIT_ORDERS)
+	public Order updateOrderFulfillerStatus(Order order, Order.FulfillStatus orderFulfillerStatus, String fullFillerComment);
```
Actually public modifier is redundant on an interface

wluyima commented on api/src/main/java/org/openmrs/Order.java:
```diff
@@ -115,6 +125,19 @@
+	 * @see org.openmrs.Order.FulfillStatus
+	 * @since 2.9.0
+	 */
+	private FulfillStatus fulfillerStatus = FulfillStatus.RECEIVED;
```
Should the name be Fulfill**er**Status

wluyima commented on api/src/main/java/org/openmrs/api/db/hibernate/ImmutableOrderInterceptor.java:
```diff
@@ -23,7 +23,7 @@
 	
 	private static final String[] MUTABLE_PROPERTY_NAMES = new String[] { "dateStopped", "voided", "dateVoided", "voidedBy",
-	        "voidReason", "patient" };
+	        "voidReason", "patient", "fulfillStatus", "fulfillerComment"};
```
Shouldn't this be fulfill**er**Status?

mogoodrich replied:
+1

wluyima commented on api/src/main/java/org/openmrs/Order.java:
```diff
@@ -54,6 +54,16 @@
+	/**
+	 * Requested by TRUNK-5400
+	 */
+	public enum FulfillStatus {
```
Shouldn't the name be Fulfill**er**Status

mogoodrich replied:
+1

wluyima commented on api/src/main/resources/liquibase-update-to-2.2.xml:
```diff
@@ -305,4 +305,25 @@
+	<changeSet id="TRUNK-5400-20180731" author="fruether">
+		<preConditions onFail="MARK_RAN">
+			<not>
+				<columnExists tableName="orders" columnName="fulfiller_status"/>
```
Can you please split these into 2 separate changesets?

wluyima commented on api/src/main/resources/liquibase-update-to-2.2.xml:
```diff
@@ -305,4 +305,25 @@
+		<comment>Adding fulfill_status and  fulfiller_comment column to provider table</comment>
+		<addColumn tableName="orders">
+			<column name="fulfiller_status" type="varchar(50)" >
+				<constraints nullable="true" />
```
I thought true is the default value anyways

wluyima commented on api/src/test/java/org/openmrs/api/OrderServiceTest.java:
```diff
@@ -1172,6 +1172,58 @@ public void saveOrder_shouldSaveARevisedOrder() {
+	 * @see OrderService#saveOrder(org.openmrs.Order, OrderContext)
+	 */
+	@Test
+	public void saveOrder_shouldSaveOrderWithFulfillerStatus() {
```
I think this test isn't required

mogoodrich replied:
Is there an existing test for testing saving an order?  I would just add setting and testing the fulfiller status there instead of creating a whole new test.

wluyima replied:
Yes it does, and that's what I was thinking too.

wluyima replied:
There is a saveOrder test that ensures that one can successfully place a new order, that's the one you need to update to set the fulfiller fields.

fruether replied:
@wluyima  The following test cases exist that create a new order and save it. I decided to add it to he bolded one. Feel free to suggest another one. But I do not in particular see a test case in OrderServiceTest that creates a order and tests if it was saved with all the according attributed

- saveOrder_shouldSetOrderTypeOfTestOrderToTestOrderIfNotSetAndConceptNotMapped
- saveOrder_shouldSetOrderTypeOfDrugOrderToDrugOrderIfNotSetAndConceptNotMappe
- **saveOrder_shouldPassForADiscontinuationOrderWithNoPreviousOrder**
- saveOrder_shouldSetConceptForDrugOrdersIfNull

wluyima commented on api/src/test/resources/org/openmrs/api/include/OrderServiceTest-discontinueAmbiguousOrderByConcept.xml:
```diff
@@ -16,5 +16,5 @@
 	<drug_order order_id="112" drug_inventory_id="3" dose="321.1" dose_units="50" as_needed="false" frequency="1" dosing_type="org.openmrs.SimpleDosingInstructions" route="22" quantity="1.0" quantity_units="51" num_refills="10" dispense_as_written="0" />
 	<orders order_id="113" order_type_id="1" previous_order_id="112" order_number="113" orderer="1" urgency="ROUTINE" order_action="REVISE" concept_id="88" creator="1" date_activated="2008-08-15 00:00:00.0" date_created="2008-08-15 00:00:00.0" voided="false" patient_id="9" uuid="fd7ec000-ce8c-4af8-b612-fe39f844cff2" care_setting="1" encounter_id="7" />
-  	<drug_order order_id="113" drug_inventory_id="3" dose="321.1" dose_units="50" as_needed="false" frequency="1" dosing_type="org.openmrs.SimpleDosingInstructions" route="22" quantity="1.0" quantity_units="51" num_refills="10" dispense_as_written="0" />
+  	<drug_order order_id="113" drug_inventory_id="3" dose="321.1" dose_units="50" as_needed="false" frequency="1" dosing_type="org.openmrs.SimpleDosingInstructions" route="22" quantity="1.0" quantity_units="51" num_refills="10" dispense_as_written="0"/>
```
Is this some stray change?

wluyima commented on api/src/main/java/org/openmrs/Order.java:
```diff
@@ -403,7 +428,7 @@ public boolean isActive() {
 	 * Convenience method to determine if the order is active as of the specified date
 	 * 
-	 * @param checkDate - the date on which to check order. if null, will use current date
+	 * @param aCheckDate - the date on which to check order. if null, will use current date
```
Are these some stray changes?

mogoodrich replied:
Agree with @wluyima , this looks like a typo (and repeated below)

wluyima commented on api/src/main/java/org/openmrs/Order.java:
```diff
@@ -115,6 +125,19 @@
+	/**
+	 * Represents the status of an order received from a filler 
+	 * @see org.openmrs.Order.FulfillStatus
+	 * @since 2.9.0
```
Shouldn't the version be 2.2.0?

fruether commented:
@wluyima @mogoodrich I tried to apply the comments you two made. I kept it with two test cases though. One for updating an order and one for creating a new one.

I am not sure for if the DB-Upgrade test is as expected

dkayiwa commented:
@fruether can you change the pull request title to match what is advised here? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

mogoodrich commented on api/src/test/java/org/openmrs/util/databasechange/Database1_9_7UpgradeIT.java:
```diff
@@ -496,6 +496,27 @@ public void shouldAddTheNecessaryPrivilegesAndAssignThemToSpecificRoles() throws
+		assertTrue(orders.size() > 0);
+		
+	}
+	
```
You don't need this here... this class was specifically for updating the 1.9.7 DB upgrade and some specific updates that happened there.  

If anything, you'd create a Database2_2_0UpgradeIT test... I think this would be a *really* good idea to have one, but, generally, I don't think we've been writing tests for all DB upgrades.  If not, I think that could be handled in a separate ticket which we could create (ie "Create Database 2.2.0 upgrade IT test").  Thoughts @dkayiwa  @wluyima ?

fruether commented:
@mogoodrich @wluyima I added the DatabaseUpgrade test. I think a close review is needed since I never created one myself.

Except for this I change following:

- I removed the 'aCheckDate' and replaced it with 'checkDate' as it was before Note: It does not reflect the parameter name now

- I need to know where I should add the test cases

wluyima commented on api/src/main/java/org/openmrs/Order.java:
```diff
@@ -115,6 +126,17 @@
 	private OrderGroup orderGroup;
 	
+	/**
+	 * Represents the status of an order received from a filler 
```
filler -> fulfiller

wluyima commented on api/src/main/java/org/openmrs/api/OrderService.java:
```diff
@@ -265,6 +265,20 @@
+	 *
+	 * @param order order whose fulFillerStatus should be changed
+	 * @param orderFulfillerStatus describes the new Order.FulfillerStatus the order should be set to
+	 * @param fullFillerComment is a string of length 1024 which describes a comment that is set while changing the FulfillerStatus               
```
I don't think it's necessary to mention the length

wluyima commented on api/src/test/java/org/openmrs/util/databasechange/Database2_2_0UpgradeIT.java:
```diff
@@ -0,0 +1,115 @@
+		String FULFILLER_STATUS = "RECEIVED";
+		String FULFILLER_COMMENT = "Comment";
+		
+		upgradeTestUtil.executeDataset(STANDARD_TEST_2_2_0_DATASET);
```
What is the use of this? My guess is that the database is already at 2.1.x version and you should only be running the  upgrade script to get it to 2.2.0

fruether replied:
@wluyima  Are you refering to this line?:
`		upgradeTestUtil.executeDataset(UPGRADE_TEST_2_1_0_TO_2_20_DATASET);
`

I am not sure about this. I copied these lines from my previous created testcase in the 1.9 integration test.
If I do not use this line I am getting the following error. That's why I was keeping it inside.

`java.lang.IllegalArgumentException: Global property not found order.drugDosingUnitsConceptUuid:`


Any suggestions? Like I said earlier. The requirement of having such an upgrade test came up but I have not so much experience in this field

wluyima commented on api/src/test/resources/org/openmrs/util/databasechange/database2_1To2_20UpgradeTest-dataSet.xml:
```diff
@@ -0,0 +1,33 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
What's the use of this file?

fruether replied:
I copied this file from the 1_9_0 Upgrade test.

It contains sample data e. .g example orders. I need this because I am updating an order. This order will be set in this file.

Not the complete dataset is necessary but as I said I copied it from the other test.

wluyima commented on api/src/test/java/org/openmrs/util/databasechange/Database2_2_0UpgradeIT.java:
```diff
@@ -0,0 +1,115 @@
+		final String insertOrderQuery = "UPDATE orders SET fulfiller_status='"+ FULFILLER_STATUS + "' , fulfiller_comment = '" +FULFILLER_COMMENT + "' where order_id = 6 ";
+		DatabaseUtil.executeSQL(connection, insertOrderQuery, false);
+		connection.commit();
+		List<Map<String, String>> orders = upgradeTestUtil.select("orders", "order_id = 6 or order_id = 7 and fulfiller_comment != ''",
```
I think you should exclude order ids from the where clause, you should select the order where status and comment match the values you set them to,

fruether replied:
Done

wluyima commented on api/src/test/java/org/openmrs/util/databasechange/Database2_2_0UpgradeIT.java:
```diff
@@ -0,0 +1,115 @@
+		List<Map<String, String>> orders = upgradeTestUtil.select("orders", "order_id = 6 or order_id = 7 and fulfiller_comment != ''",
+			"fulfiller_status", "fulfiller_comment");
+		
+		assertTrue(orders.size() > 0);
```
It more concrete to check that the size is 1 because you only inserted one if you update the select to what I suggested above

fruether replied:
Done

fruether commented on api/src/test/java/org/openmrs/util/databasechange/Database2_2_0UpgradeIT.java:
```diff
@@ -0,0 +1,115 @@
+	
+	public static final String STANDARD_TEST_2_2_0_DATASET = TEST_DATA_DIR + "standardTest-2.2.0-dataSet.xml";
+	
+	public final static String DATABASE_PATH = TEST_DATA_DIR + "openmrs-2.2.0.h2.db";
```
@wluyima 
How do I create such a .db file in the first place? I have no idea how this is done? So far I just duplicated the 1.9.0 integration test. I do no know what steps are necessary to write such a test class. I am just trying to port the 1.9.0 test as good as I can since that was required after the last PR version.

wluyima replied:
You are writing a test to demonstrate that one can upgrade from 2.1.x to 2.2.0, therefore your base dataset should be a 2.1.x dataset, the h2 documentation must have steps on how to create the .db file.

fruether replied:
I am still not sure how such a file can be created out of the Hibernate definitions and for what version this should appear. I posed regarding this concern in Talk:

https://talk.openmrs.org/t/ra-865-enable-running-openmrs-2-x-on-h2-database/2616/3

fruether commented on api/src/test/java/org/openmrs/util/databasechange/Database2_2_0UpgradeIT.java:
```diff
@@ -0,0 +1,115 @@
+	public static final String UPGRADE_TEST_2_1_0_TO_2_20_DATASET = TEST_DATA_DIR
+		+ "database2_1To2_20UpgradeTest-dataSet.xml";
+	
+	public static final String STANDARD_TEST_2_2_0_DATASET = TEST_DATA_DIR + "standardTest-2.2.0-dataSet.xml";
```
I need a datatset (see below) How it is named is not my concern. 

I was just copying the 1.9 upgrade test. That test had the exact same file except as name 1.9.0...

wluyima replied:
I think you need to name the files accordingly to make the code readable. 

The way upgrade tests are meant to be setup, the assumption is that you are coming from the last released version to the next version, in this case from 2.1.x to 2.2.0, so your base database should be at 2.1.x.

fruether replied:
I renamed it to 2.1.9 as suggested. Thank you!

wluyima commented:
I think @rkorytkowski created the .db file which I think is a binary representation, probably you just need to export the DB in binary format, let's see what @rkorytkowski has to say.

fruether commented:
@wluyima Thanks for the quick response. This would mean that I would have to run the OpenMRS Version 2.1.X (I guess 2.1.9) because there the liquidbase changeset is not applied yet. Then somehow export it as db file and load it into the UpgradeTest.

It would be quite convenient if that could be crated by a maven plugin. Unfortunately I did not find a plugin that would be capable of exporting such a file
The file is close to being readable. It contains quite a few Create statements that reflect the Liquidbase changesets 
`LIQUIBASECHANGELOG`

fruether commented:
@wluyima @rkorytkowski  @mogoodrich I removed the DatabaseUpgradeTest from this PR so that it can be merged into master. However, I would volunteer to create the DatabaseUpgradeTest for version 2 if noone else volunteers or wants to do this.

My questions regarding the H2 file could be discussed in talks:  https://talk.openmrs.org/t/ra-865-enable-running-openmrs-2-x-on-h2-database/2616/2


Is there anything else that stops this PR from being merged? I think I solved all the remaining comments from the last review. Only issue I may see is that I maybe added the test to the wrong saveOrder test case ...

wluyima commented on api/src/test/java/org/openmrs/api/OrderServiceTest.java:
```diff
@@ -2683,14 +2700,25 @@ public void getRevisionOrder_shouldNotReturnAVoidedRevisionOrder() {
 	 */
 	@Test
 	public void saveOrder_shouldPassForADiscontinuationOrderWithNoPreviousOrder() {
```
I think the correct test should be one of the those in OrderIntegrationTest, possibly `shouldPlaceADrugOrder()`

fruether commented on api/src/test/java/org/openmrs/OrderEntryIntegrationTest.java:
```diff
@@ -140,6 +140,8 @@ public void shouldPlaceATestOrder() {
 		order.setSpecimenSource(conceptService.getConcept(22));
 		order.setNumberOfRepeats(3);
+		order.setFulfillerStatus(Order.FulfillerStatus.RECEIVED);
+		order.setFulfillerComment("A comment from the filler");
```
@wluyima I added the test case to the _shouldPlaceATestOrder_ test  test method. Is this sufficient? 

I guess DrugOrder is a bit more specific than TestOrder. So I choose this method instead

wluyima replied:
This is fine too

mogoodrich commented:
Looks good, thanks for doing this @fruether , and thanks for merging @wluyima !
